src/PKSim.Core/Mappers/PathAndValueBuildingBlockMapper.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System.Collections.Generic;
+using OSPSuite.Core;
+using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Core.Domain.Formulas;
+using OSPSuite.Core.Domain.Services;
+using OSPSuite.Utility;
+using OSPSuite.Utility.Collections;
+using PKSim.Core.Model;
+
+namespace PKSim.Core.Mappers
+{
+   public interface IPathAndValueBuildingBlockMapper<in T, out TBuildingBlock> : IMapper<T, TBuildingBlock>
+   {
+   }
+
+   public abstract class PathAndValueBuildingBlockMapper<T, TBuildingBlock, TBuilder> : IPathAndValueBuildingBlockMapper<T, TBuildingBlock> where T : PKSimBuildingBlock where TBuildingBlock : PathAndValueEntityBuildingBlockFromPKSim<TBuilder> where TBuilder : PathAndValueEntity
+   {
+      protected IObjectBaseFactory _objectBaseFactory;
+      protected IEntityPathResolver _entityPathResolver;
+      protected IApplicationConfiguration _applicationConfiguration;
+      private readonly ILazyLoadTask _lazyLoadTask;
+      private readonly Cache<string, IFormula> _formulaCache = new Cache<string, IFormula>(x => x.Name);
+
+      protected PathAndValueBuildingBlockMapper(IObjectBaseFactory objectBaseFactory, IEntityPathResolver entityPathResolver, IApplicationConfiguration applicationConfiguration, ILazyLoadTask lazyLoadTask)
+      {
+         _objectBaseFactory = objectBaseFactory;
+         _entityPathResolver = entityPathResolver;
+         _applicationConfiguration = applicationConfiguration;
+         _lazyLoadTask = lazyLoadTask;
+      }
+
+      protected TBuildingBlock CreateBaseObject(T pkSimBuildingBlock)
+      {
+         var buildingBlock = _objectBaseFactory.Create<TBuildingBlock>();
+
+         buildingBlock.Name = pkSimBuildingBlock.Name;
+         buildingBlock.PKSimVersion = _applicationConfiguration.Version;
+         buildingBlock.Description = pkSimBuildingBlock.Description;
+         return buildingBlock;
+      }
+
+      private TBuilder mapBuilderParameter(IParameter parameter)
+      {
+         var builderParameter = _objectBaseFactory.Create<TBuilder>();
+
+         if (parameter.Formula != null && parameter.Formula.IsCachable())
+         {
+            if (!_formulaCache.Contains(parameter.Formula.Name))
```

Comment:
I could not understand how to use the formula factory in this more generalized case. We had previously hard-coded the category to EXPRESSION_PARAMETERS but I couldn't generalize it.

In the end, I just did my own uniqueness cache scheme

Reply:
I don't understand

Reply:
yeah but the code is not  equivalent. One you were creating anew instance, here you are adding the same reference

Reply:
plus you are not adding it to the building block FormulaCache.  they won';t be serialized

Reply:
They are copied over to the bb somewhere else